Core/BookmarksCoreDataStorage.swift

```diff
@@ -0,0 +1,762 @@
+    }
+    
+    public var topLevelBookmarksItems: [BookmarkItemManagedObject] {
+        guard let folder = cachedReadOnlyTopLevelBookmarksFolder else {
+            return []
+        }
+        return folder.children?.array as? [BookmarkItemManagedObject] ?? []
+    }
+    
+    public var favorites: [BookmarkManagedObject] {
+        return readOnlyTopLevelFavoritesItems().map {
+            if let fav = $0 as? BookmarkManagedObject {
+                return fav
+            } else {
+                fatalError("Favourites shouldn't contain folders")
+            }
+        }
+    }
+    
+    public func contains(url: URL, completion: @escaping BookmarkExistsMainThreadCompletion) {
+        containsBookmark(url: url, searchType: .bookmarksAndFavorites, completion: completion)
+    }
+    
+    public func containsBookmark(url: URL, completion: @escaping BookmarkExistsMainThreadCompletion) {
+        containsBookmark(url: url, searchType: .bookmarksOnly, completion: completion)
+    }
+    
+    public func containsFavorite(url: URL, completion: @escaping BookmarkExistsMainThreadCompletion) {
+        containsBookmark(url: url, searchType: .favoritesOnly, completion: completion)
+    }
+    
+    public func bookmark(forURL url: URL, completion: @escaping (BookmarkManagedObject?) -> Void) {
+        viewContext.perform {
+    
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.predicate = NSPredicate(format: "%K == %@ AND %K == false",
+                                                 #keyPath(BookmarkManagedObject.url),
+                                                 url as NSURL,
+                                                 #keyPath(BookmarkManagedObject.isFavorite))
+            
+            let results = try? self.viewContext.fetch(fetchRequest)
+            completion(results?.first)
+        }
+    }
+    
+    public func favorite(forURL url: URL, completion: @escaping (BookmarkManagedObject?) -> Void) {
+        viewContext.perform {
+    
+            let fetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+            fetchRequest.predicate = NSPredicate(format: "%K == %@ AND %K == true",
+                                                 #keyPath(BookmarkManagedObject.url),
+                                                 url as NSURL,
+                                                 #keyPath(BookmarkManagedObject.isFavorite))
+            
+            let results = try? self.viewContext.fetch(fetchRequest)
+            completion(results?.first)
+        }
+    }
+    
+    // Just used for favicon deletion and search
+    public func allBookmarksAndFavoritesFlat(completion: @escaping ([BookmarkManagedObject]) -> Void) {
+        viewContext.perform { [weak self] in
+        
+            let fetchRequest: NSFetchRequest<BookmarkManagedObject> = BookmarkManagedObject.fetchRequest()
```

Comment:
I'd consider adding parameter to configure whether we want to obtain bookmarks objects that are already prefetched (non-faults) to improve performance of search.

Reply:
Sorry can you clarify what you mean?

Reply:
By default `fetchRequest` does not populate content of objects from the DB - so when you iterate over the result set (e.g. in search logic) it causes N (where N is number of objects) DB queries - one for each object. Instead, you can configure [`returnsObjectsAsFaults`](https://developer.apple.com/documentation/coredata/nsfetchrequest/1506756-returnsobjectsasfaults) to make things more efficient. It would be good to have that as an option here.

Reply:
Should it be an option? Sounds like I should just disable faulting for this method tbh

Reply:
I'd say yes - to not bind this API with certain usage. But I don't have strong opinion, you could just set that property to false here to improve performance of current use cases.